---
title: "Run ESTIMATE on bulk and pseudobulk samples"
author: "Stephanie J. Spielman"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: true
    toc_depth: 3
    code_folding: hide
---

The purpose of this notebook is to run ESTIMATE on bulk (counts and TPM) and pseudobulk samples and compare results to one another.

## Setup


```{r setup}
renv::load()

library(ggplot2)
library(tidyestimate)
theme_set(theme_bw())
```

### Paths

```{r paths}
data_dir <- here::here("analysis", "estimate", "data")
sample_metadata_file <- here::here("s3_files", "scpca-sample-metadata.tsv")
library_metadata_file <- here::here("s3_files", "scpca-library-metadata.tsv")
expression_files <- list.files(
  data_dir, 
  pattern = "_expression\\.rds$"
) |>
  purrr::set_names(
    \(x) {stringr::str_split_i(x, pattern = "_", i = 1)}
  )

output_file <- here::here("analysis", "estimate", "results", "estimate_results.tsv")
```

### Read input data

```{r}
sample_metadata_df <- readr::read_tsv(sample_metadata_file, show_col_types = FALSE)
library_metadata_df <- readr::read_tsv(library_metadata_file, show_col_types = FALSE)
expression_data <- expression_files |>
  purrr::map(
    \(filename) {
      readr::read_rds(file.path(data_dir, filename))
})
```

## Run ESTIMATE

We will now run ESTIMATE on all three types of counts for samples in each project.
Note all values here are unnormalized/untransformed and represent raw counts (except TPM, but these values have not been _further_ transformed).

* bulk counts from bulk RNA-Seq samples
* bulk TPM from bulk RNA-Seq samples
* pseudobulk counts from scRNA-Seq samples (sample counts summed)

```{r, message = FALSE}
estimate_df <- expression_data |>
  # map over projects
  purrr::map(
    \(project_df_list) {
      # map over expression types
      
      purrr::map(project_df_list, \(expr_df) {
        ## Run ESTIMATE
        expr_df |> 
          dplyr::rename(hgnc_symbol = gene_symbol) |>
          tidyestimate::filter_common_genes(
            id = "hgnc_symbol", 
            tidy = TRUE, # first column contains gene symbol
            tell_missing = FALSE, 
            find_alias = TRUE
          ) |> 
          tidyestimate::estimate_score(is_affymetrix = FALSE)
      }) |> 
        purrr::list_rbind(names_to = "expression_type")
  }) |> 
  purrr::list_rbind(names_to = "project_id") |>
  dplyr::rename(sample_id = sample) |>
  # this is just the sum of scores, we won't be working with it
  dplyr::select(-estimate) |> 
  # make scores long too
  tidyr::pivot_longer(
    cols = c(stromal, immune), 
    names_to = "score_type", 
    values_to = "score"
  )
```


## Visualization


Let's begin by looking at the overall values:


```{r fig.height=4, fig.width=12}
ggplot(estimate_df) + 
  aes(x = score_type, y = score, color = expression_type) + 
  geom_jitter(
    position = position_jitterdodge(jitter.height = 0, jitter.width = 0.1), 
    size = 0.75
  ) +
  scale_color_brewer(palette = "Set2") +
  facet_wrap(vars(project_id), scales = "free_x", nrow = 1) + 
  theme(legend.position = "bottom") + 
  guides(color = guide_legend(override.aes = list(size = 2)))
```
First, what do we see for immune scores?

* For SCPCP000001 and SCPCP000002, pseudobulk scores tend to be higher than bulk scores
  * However, for SCPCP000001, we have a sort of trimodal distribution of values - it would be nice if this corresponded to diagnoses?!
* For the other three projects, pseudobulk scores appear to be somewhat lower than bulk

Second, what do we see for stromal scores?

* For SCPCP000001 and SCPCP000002, pseudobulk scores look either on par with bulk or _maybe_ a bit lower, but that is mostly in SCPCP000001 and one outlier in SCPCP000002
* For the other three projects, pseudobulk and bulk scores appear highly similar



The bulk distributions appear very similar to one another, so let's compare them directly next:



```{r fig.width=10}
expr_wide_df <- estimate_df |>
  tidyr::pivot_wider(
    names_from = expression_type, 
    values_from = score
  )

ggplot(expr_wide_df) + 
  aes(x = bulk_counts, y = bulk_tpm) + 
  geom_point() +
  geom_abline(color = "darkorange2") +
  coord_equal() +
  facet_grid(
    rows = vars(score_type), 
    cols = vars(project_id)
  ) 
```

Bulk counts do not take gene length into account, but TPM does. 
Across all projects, we see that immune scores tend to be slightly higher for TPM, and by contrast stromal scores tend to be slightly lower for TPM compared to bulk.
The effect sizes are fairly small, but the effects seem likely to be significant, which we'll check separately for `stromal` and `immune`.
Below, we fit regressions that fix the slope to 1, so the intercept will tell us the bias from `y=x`:

```{r}
stromal_df <- expr_wide_df |>
  dplyr::filter(score_type == "stromal")
summary(lm(bulk_tpm ~ 1 + offset(bulk_counts), data = stromal_df))


immune_df <- expr_wide_df |>
  dplyr::filter(score_type == "immune")
summary(lm(bulk_tpm ~ 1 + offset(bulk_counts), data = immune_df))

```

Indeed, effect sizes are not very large in the scheme of the full range of values, but they are highly significant: Compared to bulk counts, immune scores are relatively higher when calculated with TPM, and stromal scores are relatively lower whem calculated with TPM.

Let's directly compare bulk to and pseudobulk.
Since TPM and bulk counts were very similar, we'll just visualize with the bulk _counts_ (not TPM) here:


```{r fig.width=10}
compare_modalities_df <- estimate_df |>
  tidyr::pivot_wider(
    names_from = expression_type, 
    values_from = score
  ) 

ggplot(compare_modalities_df) + 
  aes(x = bulk_counts, y = pseudobulk) + 
  geom_point() +
  geom_abline(color = "darkorange2") +
  geom_smooth(method = "lm") +
  facet_grid(
    rows = vars(score_type), 
    cols = vars(project_id)
  ) +
  labs(title ="pseudobulk ~ bulk counts") +
  theme(aspect.ratio = 1)
```

* There does not appear to be a significant relationship between modalities for SCPCP000002 and SCPCP000006, for either score 
  * That said, it is worth noting that all samples' immune scores, and most samples' stromal scores, for SCPCP000002 are higher in pseudobulk than in bulk
* For SCPCP000001, immune scores tend to be higher in pseudobulk but stromal scores are about the same between modalities
* For SCPCP000006 and SCPCP0000017, immune scores tend to be lower in pseudobulk but stromal scores are about the same between modalities
* Note that projects SCPCP000001 and SCPCP000002 are single-cell, but the other three projects are single-nucleus


### SCPCP000001

In SCPCP000001, we seem to have have a trimodal distribution of pseudobulk immune scores.
Does any metadata variable separate these points?


```{r}
library_project1 <- library_metadata_df |> 
  dplyr::filter(
    scpca_project_id == "SCPCP000001", 
    seq_unit == "bulk"
  ) |> 
  dplyr::select(scpca_sample_id, technology, library_method)

project1_df <- sample_metadata_df |>
  dplyr::filter(scpca_project_id == "SCPCP000001") |>
  dplyr::inner_join(library_project1) |>
  dplyr::inner_join(compare_modalities_df, by = c("scpca_sample_id" = "sample_id")) |>
  dplyr::filter(score_type == "immune")
```

We'll first look at diagnosis:

```{r}
ggplot(project1_df) + 
  aes(x = diagnosis, y = pseudobulk, color = subdiagnosis) + 
  geom_jitter(width = 0.1) + 
  ggtitle("pseudobulk ~ diagnosis") + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1))
```

The sample size here is too small to really draw a conclusion, but it's possible that the "epithelioid" smples are coming out with the highest immune scores.

Let's look at a few other variables:


```{r fig.height=10, fig.width=10}
p1 <- ggplot(project1_df) + 
  aes(x = age, y = pseudobulk) + 
  geom_point() +
  ggtitle("pseudobulk ~ age")

p2 <- ggplot(project1_df) + 
  aes(x = technology, y = pseudobulk) + 
  geom_jitter(width = 0.1) + 
  ggtitle("pseudobulk ~ technology")

p3 <- ggplot(project1_df) + 
  aes(x = tissue_location, y = pseudobulk) + 
  geom_jitter(width = 0.1) + 
  ggtitle("pseudobulk ~ tissue_location") +
  # turn off labels for now, just to see if there's a pattern
  theme(axis.text.x = element_text(angle = 30, hjust = 1))


p4 <- ggplot(project1_df) + 
  aes(x = disease_timing, y = pseudobulk) + 
  geom_jitter(width = 0.1) + 
  ggtitle("pseudobulk ~ disease_timing") +
  # turn off labels for now, just to see if there's a pattern
  theme(axis.text.x = element_text(angle = 30, hjust = 1))


patchwork::wrap_plots(
  p1, p2, p3, p4, nrow = 2
)
```

Doesn't look like there is much of a relationship here. 
Either way, what are the negative-value samples?

```{r}
project1_df |>
  dplyr::filter(pseudobulk < 0, score_type == "immune")

```

## Export

Finally, we'll export the ESTIMATE results.


```{r}
readr::write_tsv(estimate_df, output_file)
```


## Session info

```{r}
sessionInfo()
```