---
title: "Perform GSEA"
author: "Stephanie J. Spielman"
date: "`r Sys.Date()`"
output:
  html_notebook:
    toc: true
    toc_depth: 3
    code_folding: hide
params:
  msigdbr_category: "H"
  model_expr_threshold: 0.25
  reps: 50
---

The purpose of this notebook is to run GSEA on the median of linear model residuals.
Genes that are enriched in higher or lower residual values may point to classes genes that are preferentially detected by bulk vs single-cell RNASeq technologies.

In addition, we will also run the same GSEA on shuffled gene sets (with `r params$reps`) to assess whether any results from GSEA are likely to be meaningful or noise. 


* This notebook will use the `r params$msigdbr_category` gene sets from `MSigDB` for the GSEA.
* This notebook will perform analyses on models from data filtered with this expression threshold: `r params$model_expr_threshold`.

## Setup


```{r setup}
renv::load()

library(ggplot2)
library(clusterProfiler)
theme_set(theme_bw())
set.seed(2025)
```

### Paths

```{r paths}
model_dir <- here::here("analysis", "pseudobulk-bulk-prediction", "results", "models")
ensembl_symbol_df <- readr::read_tsv(
  here::here("analysis", "bulk-deconvolution", "data", "reference", "ensembl_symbol.tsv"), 
  show_col_types = FALSE
)
```

### Functions


This chunk defines a function to create a data frame with a shuffled version of genesets.
```{r}
generate_shuffle <- function(geneset_count_df, ensembl_ids){
  shuffled_df <- geneset_count_df |>
    purrr::pmap(
    \(gs_name, n) {
      data.frame(
        gs_name = gs_name, 
        ensembl_id = sample(ensembl_ids, n) # no replacement!
      )
  }) |>
  purrr::list_rbind()
  
  # double check the length 
  stopifnot("Failed to shuffle genes" = nrow(shuffled_df) == nrow(geneset_df))
  
  return(shuffled_df)
}
```



## Read and prepare data

```{r}
# get file name strings based on model_expr_threshold
ext_str <- ifelse(params$model_expr_threshold < 0, "all-genes", glue::glue("threshold-{params$model_expr_threshold}"))
```


```{r}
model_files <- list.files(
  path = model_dir,
  full.names = TRUE,
  pattern = paste0(ext_str, "\\.rds$")
) |>
  purrr::set_names(
    \(f) stringr::str_split_i(basename(f), pattern = "_", i = 1)
  )

n_projects <- 5
stopifnot(
  "Model files not found. Check the provided model file extension." = length(model_files) == n_projects
)

ranked_residuals_list <- model_files |>
  purrr::map(
  \(f) {
    model_data <- readr::read_rds(f)
    
    model_df <- model_data |>
      purrr::pluck("data") |>
      dplyr::mutate(residuals = unname(resid(model_data$model))) |>
      # get a ranked list
      dplyr::group_by(ensembl_id) |>
      dplyr::summarize(median_res = median(residuals)) |>
      dplyr::arrange(-median_res) |>
      dplyr::pull(median_res, name = ensembl_id)
  }
)
```

## Define the gene signatures of interest


```{r}
num_genesets <- msigdbr::msigdbr_collections() |>
  dplyr::filter(gs_cat == params$msigdbr_category) |>
  dplyr::pull(num_genesets)

geneset_df <- msigdbr::msigdbr(species = "Homo sapiens", category = params$msigdbr_category)
```

The `r params$msigdbr_category` signature set contains `r num_genesets` gene sets.

## Perform GSEA

Next, we'll run GSEA on this data.


```{r}
gsea_df <- ranked_residuals_list |>
  purrr::map(
  \(ranked_genes) {
    
    gsea_result <- clusterProfiler::GSEA(
      geneList = ranked_genes, 
      TERM2GENE = dplyr::select(geneset_df, gs_name, human_ensembl_gene)
    )
    gsea_result@result |>
      dplyr::summarize(
        bulk_enriched = sum(NES > 0), 
        pseudobulk_enriched = sum(NES < 0)
      )
  }) |> purrr::list_rbind(names_to = "project_id") |>
  # format data for plotting
  tidyr::pivot_longer(
    contains("enriched"), 
    names_to = "direction", 
    values_to = "count"
  ) |>
  dplyr::mutate(
    # pseudobulk should be first
    direction = forcats::fct_rev(direction), 
    # calculate fraction version
    frac = round(count/num_genesets, 2)
  )
```



## Perform GSEA with shuffled gene sets

We'll first create `r params$reps` shuffled versions of the gene sets.

```{r}
geneset_count_df <- geneset_df |>
  dplyr::count(gs_name)

unique_ensembl_ids <- geneset_df |>
  dplyr::distinct(human_ensembl_gene) |>
  dplyr::pull(human_ensembl_gene)

shuffled_geneset_df <- 1:params$reps |>
  purrr::map(
    \(i) {generate_shuffle(geneset_count_df, unique_ensembl_ids)}
  )
```


Now, we can run GSEA on all shuffled gene set versions:

(This next chunk takes a little bit to run; we'll turn off messages and warnings since this is quite a lot of GSEAs!)

```{r, message = FALSE, warning = FALSE}
shuffled_gsea_df <- ranked_residuals_list |>
  purrr::map(
  \(ranked_genes) {
    
    purrr::map(
      shuffled_geneset_df, 
      \(shuffled_rep) {
        gsea_result <- clusterProfiler::GSEA(
          geneList = ranked_genes, 
          TERM2GENE = shuffled_rep
        )
        gsea_result@result |>
          dplyr::summarize(
            bulk_enriched = sum(NES > 0), 
            pseudobulk_enriched = sum(NES < 0)
          )
      }
    ) |> purrr::list_rbind(names_to = "rep")
    
  }) |> 
  purrr::list_rbind(names_to = "project_id") |>
  # format data for plotting
  tidyr::pivot_longer(
    contains("enriched"), 
    names_to = "direction", 
    values_to = "count"
  ) |>
  dplyr::mutate(
    # pseudobulk should be first
    direction = forcats::fct_rev(direction), 
    # calculate fraction version
    frac = count/num_genesets
  )
```

## Compare GSEA to shuffled GSEA results

Finally, we'll make plots comparing these two analyses:

* A bar plot of the number of significant pathways per project detected with GSEA
* A histogram of the number of significant pathways per project detected across the `r params$reps` replicates of shuffled GSEAs

If the values between these plots are similar, that suggests our GSEA analysis is not capturing biologically meaningful signal.
Note that these plots considers all results GSEA deemed to be significant; there is no further filtering by effect size or P-value at this time.

Again, for context, the `r params$msigdbr_category` signature category analyzed here contains `r num_genesets` total gene sets.


#### GSEA results

Each pair of bars in this plot represents a single GSEA result.
The fraction represents the number of gene sets identified as significant.

```{r, fig.width = 10, fig.height = 4}
ggplot(gsea_df) +
  aes(x = project_id, y = frac, fill = direction) + 
  geom_col(position = position_dodge(width = 0.9)) +
  geom_text(
    aes(label = frac, y = frac + 0.05), 
    position = position_dodge(width = 0.9)
  ) +
  scale_fill_brewer(palette = "Dark2") +
  ylab("Fraction of gene sets\nwith significant enrichment") +
  theme(legend.position = "bottom")
```



#### Shuffled GSEA results 

This histogram depicts results from `r params$reps` GSEAs on run on shuffled gene sets.
There are therefore `r params$reps` values in the plot for each enrichment direction. 
This is technically an "overlaid" histogram of two categories, but the two categories tend to have low variation and lie at different ends of the axes.

In this plot, the Y-axis has been scaled to 1.
So, an `(X,Y)` coordinate can be interpreted as: `Y` percent of the shuffled GSEAs identified `X` gene sets enriched in the given direction.

```{r, fig.width = 10, fig.height = 4, message = FALSE}
ggplot(shuffled_gsea_df) +
  aes(x = frac, fill = direction) + 
  geom_histogram(
    aes(y = after_stat(ndensity)), 
    position = "identity", 
    binwidth = 0.05
  ) + 
  facet_wrap(vars(project_id), nrow = 1) +
  scale_fill_brewer(palette = "Dark2") +
  ylab("Fraction of gene sets\nwith significant enrichment") +
  theme(legend.position = "bottom")
```


## Session info

```{r}
sessionInfo()
```