---
title: "Expression levels of quanTIseq tumor genes"
author: Stephanie J. Spielman
date: "`r Sys.Date()`"
params:
  permutation_reps: 1000
output:
  html_notebook:
    toc: true
    toc_depth: 3
    code_folding: show
---

## Setup

```{r setup}
renv::load()

library(ggplot2)
theme_set(theme_bw())

set.seed(2025) 
```

### Paths

Input files:

```{r}
tpm_dir <- here::here("analysis", "bulk-deconvolution", "data", "tpm")
tpm_files <- list.files(
  path = tpm_dir,
  full.names = TRUE, 
  pattern = ".rds$"
)
```

## Background

`quanTIseq` uses a signature gene list to perform deconvolution: <https://github.com/icbi-lab/quanTIseq/blob/039eb16455373915ace33b3ecede09bde697d005/quantiseq/deconvolution/TIL10_signature.txt>.
```{r}
sig_genes_df <- readr::read_tsv(
  "https://raw.githubusercontent.com/icbi-lab/quanTIseq/039eb16455373915ace33b3ecede09bde697d005/quantiseq/deconvolution/TIL10_signature.txt", 
  show_col_types = FALSE
)
signature_genes <- sig_genes_df$ID
head(signature_genes)
```

Within these signature genes, there are 17 genes which they recommend filtering out of tumor analyses based on TCGA expression.
From the [`quanTIseq` paper](https://doi.org/10.1186/s13073-019-0638-6):

> Aberrant de-methylation and sequence duplication can lead to over-expression of immune signature genes. 
> Tumor RNA-seq data can be analyzed with quanTIseq setting the `--tumor` option to `TRUE`. 
> This setting discards the signature genes whose `log2(xgl + 1)` expression in the TCGA RNA-seq data exceeds 11 TPM, which are NUPR1, CD36, CSTA, HPGD, CFB, ECM1, FCGBP, PLTP, FXYD6, HOPX, SERPING1, ENPP2, GATM, PDPN, ADAM6, FCRLA, and SLC1A3.

These are also given here: <https://github.com/icbi-lab/quanTIseq/blob/039eb16455373915ace33b3ecede09bde697d005/quantiseq/deconvolution/TIL10_TCGA_aberrant_immune_genes.txt>

```{r}
tumor_genes_df <- readr::read_csv(
  "https://raw.githubusercontent.com/icbi-lab/quanTIseq/039eb16455373915ace33b3ecede09bde697d005/quantiseq/deconvolution/TIL10_TCGA_aberrant_immune_genes.txt", 
  col_names = "gene_symbol", 
  show_col_types = FALSE
)
tumor_genes <- tumor_genes_df$gene_symbol

# Indeed, they are all in the signature:
all(tumor_genes %in% signature_genes)
```

What do these genes do? 
Aside from being immune signature genes, they are _very briefly:_

* `NUPR1`: Trans factor involved in cellular stress and tumor progression
* `CD36`: Transmembrance receptor involved in lipid uptake
* `CSTA`: Cysteine protease inhibitor 
* `HPGD`: Prostaglandin metabolism
* `CFB`: Complement factor B
* `ECM1`: Extracellular matrix
* `FCGBP`: IgGFc-binding protein, involved in tumor metastasis
* `PLTP`: TP53 target gene
* `FXYD6`: phosphohippolin; Na,K-ATPase regulator
* `HOPX`: hox gene, development
* `SERPING1`: serine protease inhibitor
* `ENPP2`: enzyme involved in lipid signaling
* `GATM`: amidinotransferase involved in creatine biosynthesis
* `PDPN`: cell migration; known oncogene
* `ADAM6`: pseudogene expressed in a lot of immune cells
* `FCRLA`: cell surface glycoprotein, involved in inflammatory response
* `SLC1A3`: glutamate transporter

On the whole, while these genes which may be associated with (adult) tumors, they are not necessarily oncogenes or causatively linked to tumor progression.

Since these genes were identified from an database of _adult_ tumor samples, it's not clear whether they have similar biological relevance to pediatric tumors. 
The purpose of this notebook is to assess whether these genes indeed have aberrantly high expression values compared to the other signature genes in each of our bulk RNA-seq samples.

## Confirm signature genes are present

Since we use gene symbol identifiers here, we should confirm that all gene symbols in the `quanTIseq` reference are indeed present in our data. 
For this, we'll check that the signature genes are all present in our mapping table which converted ensembl ids to gene symbols.

```{r}
map_table <- readr::read_tsv(
  here::here("analysis", "bulk-deconvolution", "data", "reference", "ensembl_symbol.tsv"),
  show_col_types = FALSE
)
map_symbols <- map_table$gene_symbol

all(signature_genes %in% map_symbols)
```

We are indeed missing some! Which ones?

```{r}
signature_genes[which(!(signature_genes %in% map_symbols))]
```

It turns out the reason these are "missing" is because they have been deprecated since `quanTIseq` was written, or they are aliases for the official gene symbol:

- `AKAP2` was deprecated and replaced with `PALM2AKAP2` (source: <https://www.ncbi.nlm.nih.gov/gene/11217>)
- `FAM46C` is an alias for `TENT5C` (source: <https://www.ncbi.nlm.nih.gov/gene/54855>)
- `GUCY1A3` is an alias for `GUCY1A1` (source: <https://www.ncbi.nlm.nih.gov/gene/2982>)

Are these gene symbols present in our ensembl <-> symbol mapping table?

```{r}
updated_gene_names <- c("PALM2AKAP2", "TENT5C", "GUCY1A1")
all(updated_gene_names %in% map_symbols)
```

Great, we have these other versions of the gene symbols!
We'll directly replace them in the signature list here.

```{r}
signature_genes <- dplyr::case_match(
  signature_genes, 
  "AKAP2" ~ "PALM2AKAP2", 
  "FAM46C" ~ "TENT5C", 
  "GUCY1A3" ~ "GUCY1A1", 
  .default = signature_genes
)
```

When we actually run `quanTIseq`, we'll need to make sure that we have matching gene names between our TPM matrices and their signature.

## Prepare the data

We'll read in the TPM data from all projects into a single long data frame, retaining only the signature genes with an indicator column for which are the tumor genes.

```{r}
# Create data frame of TPM values for all samples
tpm_df <- tpm_files |>
  purrr::map(readr::read_rds) |> 
  # combine into a single matrix
  purrr::reduce(cbind) |>
  as.data.frame() |>
  tibble::rownames_to_column(var = "gene_symbol") |>
  # Filter to signature genes only and add indicator column for tumor genes
  dplyr::filter(gene_symbol %in% signature_genes) |>
  dplyr::mutate(tumor_gene = gene_symbol %in% tumor_genes) |>
  # Long data frame
  tidyr::pivot_longer(
    starts_with("SCPCS"), 
    names_to = "sample_id", 
    values_to = "tpm"
  )

head(tpm_df)
```

And confirm once more that we have all the genes:

```{r}
all(signature_genes %in% tpm_df$gene_symbol)
```

## Expression distributions

First, we'll visualize distributions of TPM for tumor and non-tumor genes.


```{r fig.height=12, fig.width=16}
ggplot(tpm_df) + 
  aes(x = tumor_gene, y = tpm, color = tumor_gene) + 
  geom_jitter(width = 0.1, size = 0.25) + 
  facet_wrap(~sample_id, scales = "free_y") + 
  theme(
    axis.text = element_text(size = rel(0.8)), 
    axis.title = element_text(size = rel(0.8)), 
    legend.position = "none"
  )
```

Looking over a great many jitter plots, there does not seem to be a universal trend of abberantly high expression among tumor genes.


## Permutation tests

Next, we'll perform two flavors of permutation tests to test the hypothesis that tumor genes have higher expression compared to non-tumor genes.
For each sample, we'll perform a parametric test using the t-statistic and nonparametric test using the u-statistic (aka, wilcox), doing `r params$permutation_reps` permutations each.

We'll do these tests as one-sided to test if the tumor genes have higher expression.
Code-wise, since the reference tumor genes level here is `FALSE`, we'll specify `alternative = "less"` for the one-sided test that non-tumor genes have lower expression than do tumor genes.

```{r}
samples <- unique(tpm_df$sample_id)

perm_df <- samples |>
  purrr::map(
    \(x) {
      sample_tpm <- dplyr::filter(tpm_df, sample_id == x)
      
      wilcox_p <- wilcox.test(tpm ~ tumor_gene, alternative = "less", data = sample_tpm)$p.value
      ttest_p <- t.test(tpm ~ tumor_gene, alternative = "less", data = sample_tpm)$p.value
      
      data.frame(
        sample_id = x, 
        nonparametric = wilcox_p, 
        parametric = ttest_p
      )
    }
  ) |>
  dplyr::bind_rows() |>
  tidyr::pivot_longer(ends_with("metric"), values_to = "pvalue", names_to = "test_type")
```

We'll visualize the P-value distributions.

```{r}
ggplot(perm_df) + 
  aes(x = pvalue) + 
  geom_histogram() + 
  facet_wrap(~test_type)
```

Permutation test p-values (with no corrections for multiple testing!) are all high for both nonparametric and parametric approaches.
Therefore, our samples do not show evidence that the `quanTIseq` tumor genes (again, these are immune-related genes identified from TCGA) have aberrantly high expression compared to other signature genes.

## Conclusion

Based on visual inspection and permutation testing, we don't have strong evidence that we should use the `tumor = TRUE` setting when running `quanTIseq` on our pediatric tumor samples.

## Session info

```{r}
sessionInfo()
```